### PR TITLE
Libre2 sync raw

### PR DESCRIPTION
**Pull request by jordipbou**
Adds synchronization of Libre2 raw values with xDrip+ Sync service. 

No options added as it uses Less Common Settings > Advanced settings for Libre 2 < show Raw values in Graph to decide if sync raw values (on master) and to show them on follower.

For implementing it a new message type has been added "l2r" to xDrip+ sync sender and receiver.

### CODE REVIEW

**Review by mukich — APPROVED** [review]

**Review by Estetne — APPROVED** [review]

**Comment by tolot27** [comment]
@tzachi-dar Does the OOP2 enable raw data sync? As far as I know, this is out of scope of OOP2.

**Comment by tolot27** [comment]
@mukich and @Estetne Please do not approve pull requests until you did a careful review. We have the same discussion several times. Please have look at https://github.com/NightscoutFoundation/xDrip/pull/1388#issuecomment-828700145 and the following discussion and review to get an insight what is meant be a "careful review".

File: app/src/main/java/com/eveningoutpost/dexdrip/GcmListenerSvc.java [hunk 1]
```diff
@@ -559,6 +560,9 @@ public void run() {
                     }
                 } else if (action.equals("libreBlock")) {
                     HandleLibreBlock(payload);
+                } else if (action.equals("l2rs")) {
+                    Libre2RawValue currentRawValue = Libre2RawValue.fromJSON (payload);
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/GcmListenerSvc.java:564]: Please use consistent code formatting (no spaces before bracket of function parameter list), at least for the code you have changed.

File: app/src/main/java/com/eveningoutpost/dexdrip/LibreReceiver.java [hunk 2]
```diff
@@ -87,6 +87,10 @@ public void run() {
                                     processValues(currentRawValue, smoothingValues, context);
                                 }
                                 currentRawValue.save();
+                                // Libre2 Raw readings are also sent to Sync+ Followers if
+                                // show raw graph for Libre2 option is activated.
+                                if (Pref.getBoolean("plus_follow_master", false) && prefs.getBoolean("Libre2_showRawGraph",false))
+                                    GcmActivity.syncLibre2RawReading (currentRawValue);
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/LibreReceiver.java:93]: Please use consistent code formatting (no spaces before bracket of function parameter list), at least for the code you have changed.

File: app/src/main/java/com/eveningoutpost/dexdrip/Models/Libre2RawValue.java [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 
         import android.provider.BaseColumns;
+        import android.util.Log;
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/Models/Libre2RawValue.java:5]: Please apply code formatting to the whole file.

**Review by tolot27 — CHANGES_REQUESTED** [review]
The implementation looks good so far. Please address the code formatting comments.

How well is it tested?

@tzachi-dar Please can you review this PR as well?

**Comment by jordipbou** [comment]
As requested, format has been updated.

About testing, we use this implementation daily to monitor our daughter glucose values remotely. No issues so far.

**Comment by infossk** [comment]
We have been using our own build with this functionality for more than a month and there are no problems with it.  I would be happy to be merged.

**Comment by Estetne** [comment]
We use it for about 3 months and everything is ok

**Comment by tolot27** [comment]
@jordipbou Please resolve the merge conflict.

**Comment by tolot27** [comment]
> As requested, format has been updated.

Thanks for the update. In the future, please create one commit containing the changes instead of individual commits per file.

**Comment by jordipbou** [comment]
@tolot27 Merge conflict has been resolved.

**Review by tolot27 — APPROVED** [review]

File: app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/BgGraphBuilder.java [hunk 4]
```diff
@@ -211,7 +211,8 @@ public BgGraphBuilder(Context context, long start, long end, int numValues, bool
             loaded_start = start;
             loaded_end = end;
             bgReadings = BgReading.latestForGraph(numValues, start, end);
-            if (DexCollectionType.getDexCollectionType() == DexCollectionType.LibreReceiver)
+            if (DexCollectionType.getDexCollectionType() == DexCollectionType.LibreReceiver
+            || (DexCollectionType.getDexCollectionType () == DexCollectionType.Follower && prefs.getBoolean("Libre2_showRawGraph",false)))
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/BgGraphBuilder.java:215]: Still two minor code formatting changes from previous to your commit. Please insert a space before `false` in `prefs.getBoolean("Libre2_showRawGraph",false)` at line 215 and 1323.

File: app/src/main/java/com/eveningoutpost/dexdrip/Models/Libre2RawValue.java [hunk 5]
```diff
@@ -59,4 +66,41 @@
     public static void updateDB() {
         fixUpTable(schema, false);
     }
-}
\ No newline at end of file
+
+    public String toJSON() {
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/Models/Libre2RawValue.java:70]: Please add a unit test for at least these two functions of `Libre2RawValue`.

**Review by tolot27 — CHANGES_REQUESTED** [review]
Please add unit tests were possible. We have to increase code coverage by careful testing of the implemented methods. Currently, we are at only 5% code coverage. See https://app.codecov.io/gh/tolot27/xDrip for details.

**Comment by twinko** [comment]
is it possible to push this further?

**Pull request closed without merging**